### PR TITLE
[Merged by Bors] - feat(Algebra/Homology/ShortComplex/ExactFunctor): relate preserving exact sequences and preserving finite (co)limits/homology

### PR DESCRIPTION
**Pull request by jjaassoonn**
Let `F : C ⥤ D` be an additive functor between abelian categories:

- `Functor.preservesFiniteLimits_tfae`: the following are equivalent:
  1. for every short exact sequence `0 ⟶ A ⟶ B ⟶ C ⟶ 0`,
    `0 ⟶ F(A) ⟶ F(B) ⟶ F(C) ⟶ 0` is exact.
  2. for every exact sequence `A ⟶ B ⟶ C` where `A ⟶ B` is mono,
    `F(A) ⟶ F(B) ⟶ F(C)` is exact and `F(A) ⟶ F(B)` is mono.
  3. `F` preserves kernels.
  4. `F` preserves finite limits.

- `Functor.preservesFiniteColimits_tfae`: the following are equivalent:
  1. for every short exact sequence `0 ⟶ A ⟶ B ⟶ C ⟶ 0`,
    `F(A) ⟶ F(B) ⟶ F(C) ⟶ 0` is exact.
  2. for every exact sequence `A ⟶ B ⟶ C` where `B ⟶ C` is epi,
    `F(A) ⟶ F(B) ⟶ F(C)` is exact and `F(B) ⟶ F(C)` is epi.
  3. `F` preserves cokernels.
  4. `F` preserves finite colimits.

- `Functor.exact_tfae`: the following are equivalent:
  1. for every short exact sequence `0 ⟶ A ⟶ B ⟶ C ⟶ 0`,
    `0 ⟶ F(A) ⟶ F(B) ⟶ F(C) ⟶ 0` is exact.
  2. for every exact sequence `A ⟶ B ⟶ C`, `F(A) ⟶ F(B) ⟶ F(C)` is exact.
  3. `F` preserves homology.
  4. `F` preserves both finite limits and finite colimits.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 1]
```diff
@@ -23,6 +22,164 @@ open Limits ZeroObject
...
+    · rintro _ _ (⟨⟩ | ⟨⟩ | ⟨_⟩) <;> simp
+    · rintro _ _ (⟨⟩ | ⟨⟩ | ⟨_⟩) <;> simp
+    · ext (⟨⟩|⟨⟩) <;> simp
+    · ext (⟨⟩|⟨⟩) <;> simp
+    · exact 𝟙 _
+    · rintro (⟨⟩ | ⟨⟩) <;> simp
+    · exact 𝟙 _
+    · rintro (⟨⟩ | ⟨⟩) <;> simp
+    · ext; simp
+    · ext; simp
```
Comment (jjaassoonn) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:105]: This part is quite slow, note that `𝟙 _` cannot be put into `?_`

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 2]
```diff
@@ -46,6 +47,173 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
...
+  epi_g :=
+    letI := F.preservesEpimorphismsOfPreservesShortComplexExact h
+    letI : Epi S.g := hS.epi_g
+    map_epi _ _
+
+set_option maxHeartbeats 500000 in
+/--
+If a functor `F : C ⥤ D` preserves exact sequences, then it preserves kernels.
+-/
+noncomputable def preservesKernelsOfPreservesShortComplexExact
```
Comment (joelriou) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:98]: There is probably some range of improvements in this definition, but more generally, I would suggest proving a TFAE for an additive functor between abelian categories:

1. If `S` is *short* exact, then `S.map F` is exact and `F.map S.f` is mono;
2. If `S` is exact and `S.f` is mono, the same applies to `S.map F`.
3. `F` preserves kernels.
4. `F` preserves finite limits.

Using the dual result, we should get that `F` preserves `ShortComplex.ShortExact` iff it preserves both finite limits and finite colimits, and with some more work we would also obtain that this is equivalent to preserving `ShortComplex.Exact` (and preserving homology). (You have obtained this, but it is also an opportunity to formalise criteria for left exact and right exact functors.)

Comment (joelriou) [reply Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:98]: As I am removing the old homology API (which implies some refactor work in PR #14569), it turns out that we essentially already have the equivalences you should need for the exactness of tensor product (see files `CategoryTheory.Abelian.Exact`, `Algebra.Homology.ShortComplex.PreservesHomology` and `Algebra.Homology.ShortComplex.ExactFunctor`).

Comment (jjaassoonn) [reply Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean]: > There is probably some range of improvements in this definition, but more generally, I would suggest proving a TFAE for an additive functor between abelian categories:
> 
> 1. If `S` is _short_ exact, then `S.map F` is exact and `F.map S.f` is mono;
> 2. If `S` is exact and `S.f` is mono, the same applies to `S.map F`.
> 3. `F` preserves kernels.
> 4. `F` preserves finite limits.
> 
> Using the dual result, we should get that `F` preserves `ShortComplex.ShortExact` iff it preserves both finite limits and finite colimits, and with some more work we would also obtain that this is equivalent to preserving `ShortComplex.Exact` (and preserving homology). (You have obtained this, but it is also an opportunity to formalise criteria for left exact and right exact functors.)

This is done now! Thanks for the suggestion

Comment (joelriou) [reply Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean]: Great. But, you need to merge with master and the proofs can be improved...

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 3]
```diff
@@ -46,6 +47,244 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
...
+lemma preservesMonomorphisms_of_preserves_shortExact_left
+    (h : ∀ (S : ShortComplex C), S.ShortExact → (S.map F).Exact ∧ Mono (F.map S.f)) :
+    F.PreservesMonomorphisms where
+  preserves {X Y} f m := by
+    let S : ShortComplex C := .mk f (cokernel.π f) $ by simp
+    have e : S.ShortExact :=
+    { exact := ShortComplex.exact_of_g_is_cokernel _ $ cokernelIsCokernel _
+      mono_f := inferInstance
+      epi_g := inferInstance }
+    exact h S e |>.2
```
Comment (joelriou) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:72]: You will need to import `Mathlib.CategoryTheory.Abelian.Exact`:
```suggestion
  preserves f _ := (h _ { exact := ShortComplex.exact_cokernel f }).2
```

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 4]
```diff
@@ -46,6 +47,244 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
...
+
+lemma preserves_finite_limits_tfae : List.TFAE
```
Comment (joelriou) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:74]:
```suggestion
lemma preservesFiniteLimits_tfae : List.TFAE
```

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 5]
```diff
@@ -46,6 +47,244 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
...
+        simpa using ((kernelCompMono _ _).inv) ≫= e1)
+      mono_f := inferInstance
+      epi_g := inferInstance }
+
+    have := (s.map F).exact_and_mono_f_iff_f_is_kernel.1 (h _ se) |>.some
+    apply ShortComplex.exact_of_f_is_kernel
+    simp only [ShortComplex.map_X₂, ShortComplex.map_X₃, ShortComplex.map_g, ShortComplex.map_X₁,
+      ShortComplex.map_f] at this ⊢
+    apply isKernelCompMono (i := this) (g := F.map $ image.ι S.g)
+    simp [← F.map_comp]
```
Comment (joelriou) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:102]:
```suggestion
  · rintro hF S ⟨hS, hf⟩
    have := preservesMonomorphisms_of_preserves_shortExact_left F hF
    refine ⟨?_, inferInstance⟩
    let T := ShortComplex.mk S.f (Abelian.coimage.π S.g) (Abelian.comp_coimage_π_eq_zero S.zero)
    let φ : T.map F ⟶ S.map F :=
      { τ₁ := 𝟙 _
        τ₂ := 𝟙 _
        τ₃ := F.map (Abelian.factorThruCoimage S.g)
        comm₂₃ := by
          dsimp
          rw [Category.id_comp, ← F.map_comp, cokernel.π_desc] }
    exact (ShortComplex.exact_iff_of_epi_of_isIso_of_mono φ).1
      (hF T { exact := (S.exact_iff_exact_coimage_π).1 hS }).1
 ```

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 6]
```diff
@@ -46,6 +47,244 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
...
+    · rintro _ _ (⟨⟩ | ⟨⟩ | ⟨_⟩) <;> simp
+    · rintro _ _ (⟨⟩ | ⟨⟩ | ⟨_⟩) <;> simp
+    · ext (⟨⟩|⟨⟩) <;> simp
+    · ext (⟨⟩|⟨⟩) <;> simp
+    · exact 𝟙 _
+    · rintro (⟨⟩ | ⟨⟩) <;> simp
+    · exact 𝟙 _
+    · rintro (⟨⟩ | ⟨⟩) <;> simp
+    · ext; simp
+    · ext; simp
```
Comment (joelriou) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:134]:
```suggestion
  tfae_have 2 → 3
  · intro hF X Y f
    refine ⟨preservesLimitOfPreservesLimitCone (kernelIsKernel f) ?_⟩
    apply (KernelFork.isLimitMapConeEquiv _ F).2
    let S := ShortComplex.mk _ _ (kernel.condition f)
    let hS := hF S ⟨ShortComplex.exact_kernel f, by infer_instance⟩
    have : Mono (S.map F).f := hS.2
    exact hS.1.fIsKernel
```

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 7]
```diff
@@ -46,6 +47,244 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
...
+
+  tfae_have 3 → 4
+  · intro h; refine ⟨?_⟩
+    apply (config := {allowSynthFailures := true}) preservesFiniteLimitsOfPreservesKernels
+    exact fun {X Y} f => (h f).some
```
Comment (joelriou) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:139]:
```suggestion
  tfae_have 3 → 4
  · intro hF
    have := fun X Y (f : X ⟶ Y) ↦ (hF f).some
    exact ⟨by apply preservesFiniteLimitsOfPreservesKernels⟩
```

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 8]
```diff
@@ -46,6 +47,244 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
...
+      | .zero => Iso.refl _
+      | .one => Iso.refl _) ?_
+      · rintro (_|_) (_|_) (_|_|_) <;> simp
+    · refine ⟨?_, ?_, ?_, ?_⟩
+      · refine ⟨𝟙 _, ?_⟩
+        rintro (_|_) <;> simp [← F.map_comp]
+      · refine ⟨𝟙 _, ?_⟩
+        rintro (_|_) <;> simp [← F.map_comp]
+      · ext; simp
+      · ext; simp
```
Comment (joelriou) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:157]:
```suggestion
  tfae_have 4 → 1
  · rintro ⟨_⟩ S hS
    exact (S.map F).exact_and_mono_f_iff_f_is_kernel |>.2
      ⟨KernelFork.mapIsLimit _ hS.fIsKernel F⟩
```

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 9]
```diff
@@ -46,6 +47,244 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
...
+
+  tfae_finish
+
+open ZeroObject in
+lemma preserves_shortComplex_exact_iff_preserves_finite_limit_colimit :
```
Comment (joelriou) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:267]: Here, you should make a `preservesHomology_tfae` (or `exact_tfae`?) which would say that it is equivalent to preserve `ShortExact`, to preserve `Exact`, to `PreservesHomology` and the combination of the preservation of finite limits/colimits. You should use `preservesFiniteLimitsOfPreservesHomology`, `preservesFiniteLimitsOfPreservesHomology` (the other direction is an instance), and also `preservesHomologyOfMapExact`.

Comment (jjaassoonn) [reply Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean]: Thanks for the suggestions and the code golfing!!

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 10]
```diff
@@ -46,6 +82,204 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
...
+      ⟨KernelFork.mapIsLimit _ hS.fIsKernel F⟩
+
+  tfae_finish
+
+/--
+If a functor `F : C ⥤ D` preserves exact sequences on the right hand side (i.e.
+if `0 ⟶ A ⟶ B ⟶ C ⟶ 0` is exact then `F(A) ⟶ F(B) ⟶ F(C) ⟶ 0` is exact),
+then it preserves epimorphisms.
+-/
+lemma preservesEpimorphism_of_preserves_shortExact_right
```
Comment (joelriou) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:160]:
```suggestion
lemma preservesEpimorphisms_of_preserves_shortExact_right
```

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 11]
```diff
@@ -46,6 +82,204 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
...
+section
+
+variable {C D : Type*} [Category C] [Category D] [Abelian C] [Abelian D]
+variable (F : C ⥤ D) [F.Additive]
+
+/--
+If a functor `F : C ⥤ D` preserves short exact sequences on the left hand side, (i.e.
+if `0 ⟶ A ⟶ B ⟶ C ⟶ 0` is exact then `0 ⟶ F(A) ⟶ F(B) ⟶ F(C)` is exact)
+then it preserves
+monomorphism.
```
Comment (joelriou) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:96]:
```suggestion
then it preserves monomorphism.
```

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 12]
```diff
@@ -46,6 +82,204 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
...
+- `F` preserves exact sequences, i.e. if `A ⟶ B ⟶ C` is exact then `F(A) ⟶ F(B) ⟶ F(C)` is exact.
+- `F` preserves both finite limits and finite colimits.
+- `F` preserves homology.
+-/
+lemma exact_tfae : List.TFAE
+    [
+      ∀ (S : ShortComplex C), S.ShortExact → (S.map F).ShortExact,
+      ∀ (S : ShortComplex C), S.Exact → (S.map F).Exact,
+      Nonempty (PreservesFiniteLimits F) ∧ Nonempty (PreservesFiniteColimits F),
+      Nonempty (PreservesHomology F)
```
Comment (joelriou) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:231]: The statement would be more parallel to the previous lemmas if 3 and 4 were exchanged.

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 13]
```diff
@@ -46,6 +82,204 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
...
+      let s : ShortComplex C := .mk (0 : 0 ⟶ S.X₁) S.f $ by simp
+      exact (s.map F).exact_iff_mono (by simp) |>.1 $ hF s (s.exact_iff_mono rfl |>.2 hS2)
+    · intro S ⟨hS1, hS2⟩
+      refine ⟨hF _ hS1, ?_⟩
+      let s : ShortComplex C := .mk S.g (0 : S.X₃ ⟶ 0) $ by simp
+      exact (s.map F).exact_iff_epi (by simp) |>.1 $ hF s (s.exact_iff_epi rfl |>.2 hS2)
+
+  tfae_have 3 → 4
+  · rintro ⟨⟨_⟩, ⟨_⟩⟩
+    exact ⟨inferInstance⟩
```
Comment (joelriou) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:268]:
```suggestion
```

File: Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean [hunk 14]
```diff
@@ -46,6 +82,204 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
...
+  tfae_have 4 → 3
+  · rintro ⟨h⟩
+    exact ⟨⟨preservesFiniteLimitsOfPreservesHomology F⟩,
+      ⟨preservesFiniteColimitsOfPreservesHomology F⟩⟩
+
+  tfae_have 3 → 2
+  · rintro ⟨⟨h1⟩, ⟨h2⟩⟩
+    haveI : PreservesHomology F := inferInstance
```
Comment (joelriou) [inline Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean:276]:
```suggestion
```

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**